src/swdrv.c

```diff
@@ -152,9 +152,52 @@ s32 _swByteGet(s32 id)
     return gp->lsw[id];
 }
 
-asm UNKNOWN_FUNCTION(func_80038204)
+// Shoutouts to an anonymous decomp.me user (Grumpy Nighingale) for matching it from 95.34% to 100%
+s32 func_80038204()
 {
-    #include "asm/80038204.s"
+    CoinThing *pCoinThings; // at r31
+    char *mapname;
+    s32 i; // at r29
+    CoinThing *pCVar3;
+    CoinThing *pCVar2;
+    s32 id;
+    pCoinThings = gp->coinThings;
+
+    if (seqGetSeq() != SEQ_MAPCHANGE)
+      return -1;
+      
+    mapname = gp->mapName;
+    for (i = 0; i < MAX_COIN_MAP; i++)
+        if (strcmp(mapname, pCoinThings[i].mapName) == 0) break;
+    
+    if (i >= MAX_COIN_MAP)
+    {
+        s32 i;
+        CoinThing *pCVar2;
+        pCVar3 = gp->coinThings;
+        pCVar2 = gp->coinThings;
+        if (strncmp(mapname, pCVar2[0].mapName, 3) && strncmp(mapname, "bos", 3))
```

Comment:
Imo it's a bit clearer to write strncmp explicitly with a `== 0` or `!= 0` since it has a weird return value